Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1332 @@
+
+The important bit for us now is this:
+
+> Parsing of white space is an important part of any parser. We propose a convention where **every lexeme parser assumes no spaces before the lexeme and consumes all spaces after the lexeme**; this is what the [`lexeme`](https://hackage.haskell.org/package/megaparsec-9.2.0/docs/Text-Megaparsec-Char-Lexer.html#v:lexeme) combinator does, and so it's enough to wrap every [`lexeme`](https://hackage.haskell.org/package/megaparsec-9.2.0/docs/Text-Megaparsec-Char-Lexer.html#v:lexeme) parser with lexeme to achieve this. Note that you'll need to call [`space`](https://hackage.haskell.org/package/megaparsec-9.2.0/docs/Text-Megaparsec-Char-Lexer.html#v:space) manually to consume any white space before the first lexeme (i.e. at the beginning of the file).
+
+#### S-expressions: The elegant way
+
+The first thing you should do is import the appropriate lexing module. Megaparsec recommends to import this module qualified.
+
+```hs
+import qualified Text.Megaparsec.Char.Lexer as L
+```
+
+Now we need to define what is whitespace. Until now, with the usage of `space`, we have considered that it may be a space, tab, newline, carriage return, form feed, or vertical tab. Megaparsec can go beyond, and also consider line comments (such as `-- comment`) and block comments (such as `{- comment -}` as whitespace. Since we get those for free, what about using `;;` for line comments and `/* */` for block comments?
+
+Begin by defining `skipSpace`. We use the `space` helper function from the lexer module, which expects a function to parse 1 or more spaces, a line comment, and a block comment.
+
+```hs
+skipSpace :: Parser ()
+skipSpace = L.space
+  -- Like `space`, but skips 1 or more space characters.
+  space1
+  -- Skip from ;; until a newline.
+  (L.skipLineComment ";;")
+  -- Skip from /* until */. There is also `skipBlockComment`, but it doesn't handle nested comments.
+  (L.skipBlockCommentNested "/*" "*/")
+```
+
+And now define a `lexeme` parser, which indicates how to consume space after the given parser.
+
+```hs
+lexeme :: Parser a -> Parser a
+lexeme = L.lexeme skipSpace
+```
+
+Now you can use `lexeme` in each of the parsers, and they will automatically skip whitespace after their definitions. Note that we also use `lexeme` before our first opening parenthesis, so we can skip space just after the `(`.
+
+```hs
+sexp :: Parser (SExp, [SExp])
+sexp = label "S-expression" $ lexeme $
+  between (lexeme (char '(')) (char ')') ((,) <$> atom <*> many atom)
+```
+
+You may add `lexeme` to other parsers where it's desirable to skip whitespace after the rule. For instance:
+
+```hs
+integer :: Parser Integer
+integer = label "integer" $ lexeme $ read <$> some numberChar
+```
+
+**Note**: The reader should do the same for `str`, `bool`, and `identifier`.
```

Comment:
... or on whole atom instead, to not repeat themselves.

Reply:
Ah, nice observation! I didn't think of adding it to `atom`. I added a comment about it.